src/data/fin.lean

```diff
@@ -1012,6 +1012,168 @@ def append {α : Type*} {o : ℕ} (ho : o = m + n) (u : fin m → α) (v : fin n
   then u ⟨i, h⟩
   else v ⟨(i : ℕ) - m, (nat.sub_lt_left_iff_lt_add (le_of_not_lt h)).2 (ho ▸ i.property)⟩
 
+variables {β : Type*} {o : ℕ} {u : fin m → β} {v : fin n → β}
+
+lemma append_apply_left (ho : o = m + n) (k : fin o) (h : (k : ℕ) < m) :
+  append ho u v k = u (k.cast_lt h) :=
+dif_pos h
+
+@[simp] lemma append_apply_left_cast (ho : o = m + n) (k : fin m) :
+  append ho u v (cast ho.symm (cast_add n k)) = u k :=
+(append_apply_left ho (cast ho.symm (cast_add n k)) k.2).trans $
+  congr_arg _ (fin.ext rfl)
+
+lemma append_apply_right (ho : o = m + n) (k : fin o) (h : ¬(k : ℕ) < m) :
+  append ho u v k = v ⟨(k : ℕ) - m, (nat.sub_lt_left_iff_lt_add $ le_of_not_lt h).2 (ho ▸ k.2)⟩ :=
+dif_neg h
+
+@[simp] lemma append_apply_right_cast (ho : o = m + n) (k : fin n) :
+  append ho u v (cast ho.symm (nat_add m k)) = v k :=
+(append_apply_right ho (cast ho.symm (nat_add m k)) (not_lt.2 $ le_add_right (le_refl _))).trans $
+  congr_arg _ (fin.ext $ nat.add_sub_cancel_left _ _)
+
+lemma append_nil {d : fin 0 → β} (ho : o = m) {x : fin o} :
+  @append 0 m _ o ho u d x = u (cast ho x) :=
+(@append_apply_left 0 m _ o _ _ ho x (ho ▸ x.2)).trans $ by congr
+
+@[simp] lemma append_nil' {d : fin 0 → β} :
+  append rfl u d = u :=
+funext $ λ x, (append_nil _).trans (congr_arg _ $ fin.ext rfl)
+
+lemma nil_append {d : fin 0 → β} (ho : o = m) {i : fin o} :
+  append (show o = 0 + m, by rwa zero_add) d u i = u (cast ho i) :=
+(append_apply_right _ _ (nat.not_lt_zero _)).trans $ by congr
+
+@[simp] lemma nil_append' {d : fin 0 → β} :
+  append (zero_add _).symm d u = u :=
+funext $ λ x, (nil_append rfl).trans (congr_arg _ $ fin.ext rfl)
+
+lemma append_one_eq_cons (ho : o = 1 + m) {x : β} (i : fin o) :
```

Comment:
The choice of implicit/explicit arguments in these lemmas seems a little arbitrary to me. Is there reasoning behind it? For example, the implicit `x` here seems analogous to the explicit `x` in the following lemma. Above, you arguably could make `k` implicit when it's followed by `h`.